identity-store/persistence/cqldb/persist.go

```diff
@@ -0,0 +1,227 @@
+package cqldb
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/gocql/gocql"
+	"github.com/plgd-dev/hub/v2/identity-store/persistence"
+	"github.com/plgd-dev/hub/v2/pkg/cqldb"
+)
+
+// PersistenceTx prevents data race for a sequence of read and write operations.
+type PersistenceTx struct {
+	tx    *gocql.Session
+	table string
+	err   error
+	ctx   context.Context
+}
+
+// NewTransaction creates a new transaction.
+// A transaction must always be closed:
+//
+//	tx := s.persistence.NewTransaction()
+//	defer tx.Close()
+func (s *Store) NewTransaction(ctx context.Context) persistence.PersistenceTx {
+	return &PersistenceTx{tx: s.client.Session(), table: s.Table(), err: nil, ctx: ctx}
+}
+
+// Retrieve device's authorization details.
+func (p *PersistenceTx) Retrieve(deviceID, owner string) (_ *persistence.AuthorizedDevice, ok bool, err error) {
+	if p.err != nil {
+		return nil, false, p.err
+	}
+
+	var b strings.Builder
+	b.WriteString(cqldb.SelectCommand + " ")
+	b.WriteString(ownerKey)
+	b.WriteString(",")
+	b.WriteString(deviceIDKey)
+	b.WriteString(" " + cqldb.FromClause + " ")
+	b.WriteString(p.table)
+	b.WriteString(" " + cqldb.WhereClause + " ")
+	b.WriteString(ownerKey)
+	b.WriteString("='")
+	b.WriteString(owner)
+	b.WriteString("' AND ")
+	b.WriteString(deviceIDKey)
+	b.WriteString("=")
+	b.WriteString(deviceID)
+
+	var d persistence.AuthorizedDevice
+	err = p.tx.Query(b.String()).WithContext(p.ctx).Scan(&d.Owner, &d.DeviceID)
+	if err != nil && errors.Is(err, gocql.ErrNotFound) {
+		return nil, false, nil
+	}
+	if err != nil {
+		return nil, false, err
+	}
+	return &d, true, nil
+}
+
+// RetrieveByDevice device's authorization details.
+func (p *PersistenceTx) RetrieveByDevice(deviceID string) (_ *persistence.AuthorizedDevice, ok bool, err error) {
+	if p.err != nil {
+		err = p.err
+		return
+	}
+
+	var b strings.Builder
+	b.WriteString(cqldb.SelectCommand + " ")
+	b.WriteString(ownerKey)
+	b.WriteString(",")
+	b.WriteString(deviceIDKey)
+	b.WriteString(" " + cqldb.FromClause + " ")
+	b.WriteString(p.table)
+	b.WriteString(" " + cqldb.WhereClause + " ")
+	b.WriteString(deviceIDKey)
+	b.WriteString("=")
+	b.WriteString(deviceID)
```

Comment:
SC reports duplication for `Retrieve` and `RetrieveByDevice`. It can be solved by a simple helper method:

```go

func (p *PersistenceTx) retrieveDeviceByQuery(query string) (_ *persistence.AuthorizedDevice, ok bool, err error) {
var d persistence.AuthorizedDevice
	err = p.tx.Query(query).WithContext(p.ctx).Scan(&d.Owner, &d.DeviceID)
	if err != nil && errors.Is(err, gocql.ErrNotFound) {
		return nil, false, nil
	}
	if err != nil {
		return nil, false, err
	}
	return &d, true, nil
}

func (p *PersistenceTx) Retrieve(deviceID, owner string) (_ *persistence.AuthorizedDevice, ok bool, err error) {
  ...
  build query
  ...

  return p.retrieveDeviceByQuery(b.String());
}

func (p *PersistenceTx) RetrieveByDevice(deviceID string) (_
  ...
  build query
  ...
  
  return p.retrieveDeviceByQuery(b.String());
}

```

Reply:
Fixed